apps/designer/app/shared/tree-utils/move-instance.ts

```diff
@@ -0,0 +1,59 @@
+import { type Instance } from "@webstudio-is/react-sdk";
+import { insertInstanceMutable } from "./insert-instance";
+
+type InstanceMoveSpec = {
+  id: Instance["id"];
+  newParentId: Instance["id"];
+  newIndex: number;
+};
+
+const extractInstance = (
+  treeNode: Instance,
+  id: Instance["id"]
+): Instance | null => {
+  const index = treeNode.children.findIndex(
+    (child) => typeof child !== "string" && child.id === id
+  );
+
+  if (index !== -1) {
+    const instance = treeNode.children[index];
+
+    // for TypeScript
+    if (typeof instance === "string") {
+      throw new Error("Cannot happen");
+    }
+
+    treeNode.children.splice(index, 1);
+    return instance;
+  }
+
+  for (const child of treeNode.children) {
+    if (typeof child !== "string") {
+      const extracted = extractInstance(child, id);
+      if (extracted !== null) {
+        return extracted;
+      }
+    }
+  }
+
+  return null;
+};
+
+export const moveInstanceMutable = (
```

Comment:
this looks like what is usually called "parenting", which we already do in useSubscribeReparentInstance: basically its 2 calls: deleteInstanceMutable() and insertInstanceMutable()

Reply:
also note that we use mutating functions because we call them inside a createTransaction that uses immer.js data structures, so the mutable functions don't actually mutate a datastructure that will be passed anywhere else, they mutate so that immer can record the changes and create a new immutable data structure